src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -136,6 +145,7 @@ void MergeTreeBackgroundExecutor<Queue>::routine(TaskRuntimeDataPtr item)
         /// So, the destructor of a task and the destructor of a storage will be executed concurrently.
         item->task.reset();
         item->is_done.set();
+        item = nullptr;
```

Comment:
AFAICS the problem is that lambda `erase_from_active` captures the `item` and so it will increment the refcnt, and eventually it will be freed when it will destroying object on function exit, however I don't see how `item = nullptr` will help here (and also `std::move(item)`)

Yes, it will work right now correctly because you added `item.reset()` in the `removeTasksCorrespondingToStorage`, but maybe it is better to fix the lambda capture list instead?